applications/AdjointFluidApplication/custom_utilities/numerical_diffusion.h

```diff
@@ -0,0 +1,377 @@
+        array_1d< double, TNumNodes > N;
+        double Volume;
+
+        GeometryType& rGeom = pCurrentElement->GetGeometry();
+
+        GeometryUtils::CalculateGeometryData(rGeom,DN_DX,N,Volume);
+
+        MatrixType velocity_gradient;
+
+        CalculateVelocityGradientTensor<TDim>(pCurrentElement, rCurrentProcessInfo, velocity_gradient);
+
+        double velocity_divergence = 0.0;
+        for (IndexType i=0;i<TDim; i++)
+            velocity_divergence += velocity_gradient(i,i);
+        
+        Eigen::Matrix<double, TDim+1, TDim+1>  characteristic_matrix;
+        for (IndexType i=0; i < TDim; i++)
+            characteristic_matrix(i,i) = 0.5 * velocity_divergence - velocity_gradient(i,i);
+        for (IndexType i=0; i < TDim; i++)
+            for (IndexType j=i+1; j < TDim; j++)
+            {
+                characteristic_matrix(i,j) =  velocity_gradient(i,j);
+                characteristic_matrix(j,i) =  velocity_gradient(j,i);
+            }
+        for (IndexType i=0; i < TDim + 1; i++)
+        {
+            characteristic_matrix(TDim, i) = 0.0;
+            characteristic_matrix(i, TDim) = 0.0;
+        }
+        characteristic_matrix(TDim, TDim) = 0.5 * velocity_divergence;
+
+        Eigen::JacobiSVD<Eigen::Matrix<double, TDim+1, TDim+1>> svd(
+                characteristic_matrix, Eigen::ComputeThinU | Eigen::ComputeThinV
+                );
+
+        const auto& S = svd.singularValues();
+
+        double numerical_viscosity;
+        numerical_viscosity = S[0];
+
+        return numerical_viscosity;
+    }
+
+    template<unsigned int TDim>
+    double CalculateNumericalDiffusionSVMethodPressureDecoupled(Element::Pointer pCurrentElement, const ProcessInfo& rCurrentProcessInfo)
+    {
+        constexpr unsigned int TNumNodes = TDim + 1;
+
+        boost::numeric::ublas::bounded_matrix<double, TNumNodes, TDim> DN_DX;
+        array_1d< double, TNumNodes > N;
+        double Volume;
+
+        GeometryType& rGeom = pCurrentElement->GetGeometry();
+
+        GeometryUtils::CalculateGeometryData(rGeom,DN_DX,N,Volume);
+
+        MatrixType velocity_gradient;
+        CalculateVelocityGradientTensor<TDim>(pCurrentElement, rCurrentProcessInfo, velocity_gradient);
+
+        double velocity_divergence = 0.0;
+        for (IndexType i=0;i<TDim; i++)
+            velocity_divergence += velocity_gradient(i,i);
+        
+        Eigen::Matrix<double, TDim+1, TDim+1>  characteristic_matrix;
+        for (IndexType i=0; i < TDim; i++)
+            characteristic_matrix(i,i) = 0.5 * velocity_divergence - velocity_gradient(i,i);
+        for (IndexType i=0; i < TDim; i++)
+            for (IndexType j=i+1; j < TDim; j++)
+            {
+                characteristic_matrix(i,j) =  velocity_gradient(i,j);
+                characteristic_matrix(j,i) =  velocity_gradient(j,i);
+            }
+
+        Eigen::JacobiSVD<Eigen::Matrix<double, TDim+1, TDim+1>> svd(
```

Comment:
Eigen can only be used through the EigenSolversApplication, and it must be strictly optional